tests/specs/lemmas.k

```diff
@@ -17,9 +17,12 @@ module LEMMAS
 
     // chop range & simplification
     rule chop(I) => I requires #rangeUInt( 256 , I ) [simplification]
-    rule 0 <=Int chop(_V)             => true        [simplification]
+    rule 0 <=Int chop(_V)             => true        [simplification, smt-lemma]
     rule         chop(_V) <Int pow256 => true        [simplification]
 
+    rule chop ( ( maxUInt48 &Int X:Int ) *Int Y:Int ) => ( maxUInt48 &Int X:Int ) *Int Y
+      requires 0 <=Int X andBool 0 <=Int Y andBool Y <=Int pow208 [simplification]
+
```

Comment:
If I understand correctly, the maximum integer `(( maxUInt48 &Int X:Int ) *Int Y:Int )` could be `2^256 - 2^208`. Is this correct, and what you saw too? It seems like a strange bound. The maximum integer for this simplification to be safe is `2^256 - 1`.

Reply:
Yes, that's correct, but those were the constraints from the test on `X` and `Y`that resulted in this simplification - actually, it was with `0 <= X < pow208` instead of the double `<=`, so I might change that bit, but it is actually safe as-is. The bound on `Y` can be increased maximally to, I believe, `floor(pow256 / (pow48 - 1))` but that is quite an unpleasant number to deal with in the reasoning.

Reply:
For rules like this we can do with a set of simpler range lemmas like this (that allow the `chop` to simplify out with it's relevant lemmas):

```
rule 0 <=Int ((maxUInt48 &Int X) *Int Y) => true requires 0 <=Int X andBool 0 <=Int Y [simplification]
rule ((maxUInt48 &Int X) *Int Y) <Int pow256 => true requires 0 <=Int X andBool Y <Int pow208 [simplification]
```

These rules may be helpful in other contexts too, not sure. But it's a different way to say it, which allows re-using other existing reasoning. It maybe be that only one of these lemmas is needed ^^^

Reply:
That does sound like the way to go, but when I test it by putting

```
    claim <k> chop ( ( maxUInt48 &Int X:Int ) *Int Y:Int ) => ( maxUInt48 &Int X ) *Int Y ... </k>
      requires 0 <=Int X andBool 0 <=Int Y andBool Y <Int pow208
```

in `tests/specs/mcd/functional-spec.k` and 

```
    // chop range & simplification
    rule chop(I) => I requires #rangeUInt( 256 , I ) [simplification]
    rule 0 <=Int chop(_V)             => true        [simplification, smt-lemma]
    rule         chop(_V) <Int pow256 => true        [simplification]

    rule 0 <=Int ((maxUInt48 &Int X) *Int Y)     => true requires 0 <=Int X andBool 0 <=Int Y     [simplification]
    rule ((maxUInt48 &Int X) *Int Y) <Int pow256 => true requires 0 <=Int X andBool Y <Int pow208 [simplification]
```

in `tests/specs/lemmas.k`, the implication check fails, the simplifications don't get applied.